src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -77,7 +79,28 @@ private static async Task WriteMessageCoreAsyncAwaited(this PipeWriter pipeWrite
 
         private static Task WriteMessageCoreAsync(this PipeWriter pipeWriter, byte[] messageData, HttpContextServerCallContext serverCallContext, bool flush)
         {
-            WriteHeader(pipeWriter, messageData.Length);
+            var canCompress = ((serverCallContext.WriteOptions?.Flags ?? default) & WriteFlags.NoCompress) != WriteFlags.NoCompress;
+
+            // If compression has not been disabled by WriteFlags.NoCompress then
+            //   1. Use compression algorithm configured on server
+            //   2. Use compression algorithm of request
+            var compressionEncoding = canCompress
+                ? serverCallContext.GetGrpcEncoding()
+                : null;
+
+            if (compressionEncoding != null)
+            {
+                if (!GrpcProtocolHelpers.TryCompressMessage(compressionEncoding, messageData, out var compressedData))
+                {
+                    // It is not an error if the server doesn't support the incoming compression algorithm
```

Comment:
cc @jtattermusch ?

Reply:
The comment needs to be updated and logic revisited?

Reply:
So this happens when the compression encoding is set on the server for which a provider cannot be found. This would be an user error so we should just log a warning and send the uncompressed message.

For example, this would occur if the server has `gzip` provider added but the ResponseCompressionAlgorithm on the GrpcServiceOption is set to `deflate`.

Reply:
Actually I think we have this check in the ServerCallHandlerFactory, so this case can't happen.

Reply:
Correct. This can be simplified.